Dataset object for v1

ipcamit opened the pull request:
## Summary

Include a summary of major changes in bullet points:

* Supports ase-io for inputting multiple xyz configs per file 
* Supports colabfit for io directly from mongo-db instance
* Minor updates regarding file io from softlinked files
* Test for ase-io

## Additional dependencies introduced (if any)

* ASE: for file io

mjwen commented on kliff/dataset/dataset.py:
```diff
@@ -320,17 +510,63 @@ def __init__(
         path: Optional[Path] = None,
         weight: Optional[Weight] = None,
-        file_format="xyz",
+        file_format: str = "xyz",
```
Now that we have multiple ways to create a dataset, It would be good to do it using individual staticmethod, instead of the if else statement. The major drawback of this if else statement approach is the mix of incompatible arguments together. What happens if a user passes `path` and `colabfit_dataset` together? We will have to do checks in __init__. 

If would propose the below:
```python
def __init__(self, configurations: list[Configuration]):
    self.configs = configurations

@staticmethod
def from_files(cls, path, <other args>):
     # read configs from the file(s) and assign to self.configs

@staticmethod
def from_colabfit(cls, database, <other args>):
     # query configs and assign to self.configs
```

Then the `add_configs` function needs to be updated as well, probably split into two? `add_from_files` and `add_from_colabfit`? 

The actual code to read the files and query colabfit can be shared, e.g. between `from_files` and `add_from_files`.

ipcamit replied:
Added new methods, `from_*` and `add_from_*` for colabfit, path, and ase based datasources. See comment below.

mjwen reviewed (changes requested):
@ipcamit I've left some comments. Would be happy to discuss them.

ipcamit commented on kliff/dataset/dataset.py:
```diff
@@ -138,6 +169,122 @@ def to_file(self, filename: Path, file_format: str = "xyz"):
+             weight:
+        """
+        try:
+            fetched_configuration: "ColabfitConfiguration" = (
```
No. `colabfit.tools` also have a class called `Configuration`, so to disambiguate, I imported it as `ColabfitConfiguration`

mjwen replied:
OK, then let's annotate it with the real class  `ColabfitConfiguration` but not the string.

ipcamit replied:
It was annotated using typing [annotation](https://docs.python.org/3.9/library/typing.html) strings and `TYPE_CHECKING` condition, as otherwise colabfit will become a runtime dependency and raise error if colabfit-tools is not installed. I am not sure that many users will use colabfit to justify making it a hard dependency. Therefore "ColabfitConfiguration" is a string. If you want to keep style consistent etc, Then I would recommend either removing type annotations to ColabfitConfiguration completely (preferable) or creating a dummy object.

mjwen replied:
OK, then using a string annotation makes sense.

ipcamit replied:
OK. Keeping the string based annotations for now.

ipcamit commented on kliff/dataset/dataset.py:
```diff
@@ -138,6 +169,122 @@ def to_file(self, filename: Path, file_format: str = "xyz"):
+    @classmethod
+    def from_colabfit_dataobjects(
+        cls,
+        database_client: "MongoDatabase",
```
I don't understand this part. The `MongoDatabase` is class name in `colabfit.tools`, although it inherits from `MongoClient`, it is specifically `colabfit.tools.database.MongoDatase` that KLIFF wants to use. Any particular benefit of using the base class?

mjwen replied:
Ah, I misunderstood it. I did not know that colabfit inherited the Client. Then let's annotate with the real `MongoDatabase` from colabfit, but not the string.

mjwen commented on kliff/dataset/dataset.py:
```diff
@@ -301,6 +467,30 @@ def order_by_species(self):
 
+    @staticmethod
+    def _get_colabfit_property(
+        database_client: Union["MongoDatabase"],
```
Not the pymongo client, but the MongoDatabase class from colabfit.

ipcamit replied:
It was union because I though early on that colabfit will have different backends, like sql databases etc. Removed the Union.

ipcamit commented:
Pushing new changes as per suggestions:

1. make ASE import direct, and not conditional
2. Fixed missing docs
3. Changed Configuration to cls in classmethod
4. Now that ase is a dependency, moved string based typehints to Atoms
5. Added stress setter
6 Added metadata field to Configuration for colabfit dataset properties
7. **New dataset initialization methods**
8. fixed tests to reflect new dataset interface

On new dataset initialization:
Now `Dataset` object supports three `classmethods`, `from_colabfit`, `from_path`, and `from_ase`, to initialize the dataset. Similarly, it has three additional methods called `add_from_colabfit`, `add_from_path`, and `add_from_ase`, to extend the datasets from these sources.

The main difference between `*from_path` and `*from_ase` is that ASE function can also accept list of `ase.Atoms` object to initialize the dataset. Another major difference is that while ase accepts Voigt notation in stress, it usually expects the xyz file to have full 9-member matrix. Therefore we have two functions, `from_path`, and `from_ase`, where `from_ase` will always stick to `ase.io.read` function.

ipcamit commented on kliff/dataset/dataset.py:
```diff
@@ -10,8 +10,31 @@
+    colabfit_installed = False
+
+# check is ase is installed
+try:
```
ASE is now a core dependency, hence removing the checks

ipcamit commented on kliff/dataset/dataset.py:
```diff
@@ -138,6 +169,122 @@ def to_file(self, filename: Path, file_format: str = "xyz"):
             )
 
+    @classmethod
+    def from_colabfit_dataobjects(
```
Renamed the function

ipcamit commented on kliff/dataset/dataset.py:
```diff
@@ -138,6 +169,122 @@ def to_file(self, filename: Path, file_format: str = "xyz"):
+
+        Args:
+            database_client: Instance of connected MongoDatabase client, which can be used to
+            fetch database from colabfit-tools dataset.
```
Edited the documentation to reflect the common standard, and fixed missing doc fields

ipcamit commented on kliff/dataset/dataset.py:
```diff
@@ -138,6 +169,122 @@ def to_file(self, filename: Path, file_format: str = "xyz"):
+            property_ids: ID of the property instance to be associated with current configuration.
+            Usually properties would be trained against. Each associated property "field" will be
+            matched against provided list of aux_property_fields.
+             weight:
```
Filled missing docs

ipcamit commented on kliff/dataset/dataset.py:
```diff
@@ -138,6 +169,122 @@ def to_file(self, filename: Path, file_format: str = "xyz"):
+        PBC = fetched_configuration.pbc
+
+        # get energy, forces, stresses from the property ids
+        energy = Configuration._get_colabfit_property(
```
Done.

ipcamit commented on kliff/dataset/dataset.py:
```diff
@@ -138,6 +169,122 @@ def to_file(self, filename: Path, file_format: str = "xyz"):
+            atoms: ase.Atoms object.
+            weight: an instance that computes the weight of the configuration in the loss
+                function.
+            forces_key:
```
Updated docs

ipcamit commented on kliff/dataset/dataset.py:
```diff
@@ -226,6 +381,17 @@ def identifier(self, identifier: str):
+    def fingerprint(self, fingerprint):
+        self._fingerprint = fingerprint
+
+    def set_fingerprint(self, fingerprint):
```
Done.

ipcamit commented on kliff/dataset/dataset.py:
```diff
@@ -185,6 +336,10 @@ def forces(self) -> np.ndarray:
+        self._forces = forces
+
     @property
     def stress(self) -> List[float]:
```
Done.

ipcamit commented on kliff/dataset/dataset.py:
```diff
@@ -301,6 +467,30 @@ def order_by_species(self):
+            property_type: type of property to fetch
+
+        Returns: fetched value, None if query comes empty
+
```
Removed line after "Reurns"

ipcamit commented on kliff/dataset/dataset.py:
```diff
@@ -50,6 +73,9 @@ def __init__(
         stress: Optional[List[float]] = None,
         weight: Optional[Weight] = None,
         identifier: Optional[Union[str, Path]] = None,
+        database_client: Optional[Union["MongoDatabase"]] = None,
```
Added metadata class

ipcamit commented on kliff/dataset/dataset.py:
```diff
@@ -320,17 +510,63 @@ def __init__(
         weight: Optional[Weight] = None,
-        file_format="xyz",
+        file_format: str = "xyz",
+        colabfit_database: str = None,
```
Done

mjwen commented on kliff/dataset/dataset.py:
```diff
@@ -301,67 +499,207 @@ def order_by_species(self):
+        if colabfit_database is not None:
+            if colabfit_dataset is not None:
+                # open link to the mongo
+                if colabfit_installed:
```
We can simplify this `if colabfit_installed ` check via `monty` (an existing dependency of kliff). See https://github.com/materialsvirtuallab/monty/blob/5d37679e1f0cd58a8d165d76f3a095cb71266831/monty/dev.py#L57

ipcamit replied:
That is quite handy, done. Will use this for future as well.

mjwen commented on kliff/dataset/dataset.py:
```diff
@@ -301,67 +499,207 @@ def order_by_species(self):
+
         """
-        return self.configs
+        if colabfit_database is not None:
```
We should remove the two checks `colabfit_database` and `colabfit_dataset`? If they are called, both values must be provided.

ipcamit replied:
Removed the check, as colabfit_dataset and database now will not be none, and a positional arguments.

mjwen commented on kliff/dataset/dataset.py:
```diff
@@ -301,67 +499,207 @@ def order_by_species(self):
+    @classmethod
+    def from_path(
+        cls,
+        path: Optional[Path] = None,
```
`path` should not be `None`, either `Optional`

ipcamit replied:
`path` is now a positional argument. Converted from `Optional[Path]` to `Union[Path, str]`

mjwen commented on kliff/dataset/dataset.py:
```diff
@@ -394,11 +732,220 @@ def _read(path: Path, weight: Optional[Weight] = None, file_format: str = "xyz")
+        file_format: str = "xyz",
+    ) -> "Dataset":
+        """
+        Read configurations from ase.Atoms object and initialize a dataset. If the
```
More explanation should be added on how to use it, particularly on the fact that a user is expected to provide either `path` or `ase_atoms_list`. Same for the `add_from_ase` function.

ipcamit replied:
Added example and more explanation in docstring

mjwen commented on kliff/dataset/dataset.py:
```diff
@@ -394,11 +732,220 @@ def _read(path: Path, weight: Optional[Weight] = None, file_format: str = "xyz")
+        """
+        return self.configs
+
+    def get_num_configs(self) -> int:
```
Given that we have `__len__` (and it is preferred), lets remove this?

mjwen commented on kliff/dataset/dataset.py:
```diff
@@ -394,11 +732,220 @@ def _read(path: Path, weight: Optional[Weight] = None, file_format: str = "xyz")
+        """
+        Get the configurations.
+        """
+        return self.configs
```
Make a shallow copy (e.g. `return self.configs[:] `) to avoid modifying the internal data.

ipcamit replied:
Done.

mjwen reviewed (commented):
Thanks @ipcamit. The updates look great. I just have a few minor comments. After this, I think we should be able to merge it.

ipcamit commented on kliff/dataset/dataset.py:
```diff
@@ -301,67 +499,207 @@ def order_by_species(self):
+    @classmethod
+    def from_colabfit(
+        cls,
+        colabfit_database: str = None,
```
Yes, I guess colabfit_database and datasets should be positional argument. Will update to reflect the same

ipcamit commented on kliff/dataset/dataset.py:
```diff
@@ -301,67 +499,207 @@ def order_by_species(self):
+
+    def add_from_colabfit(
+        self,
+        colabfit_database: str = None,
```
Yes. Fixed, same as above

ipcamit commented on kliff/dataset/dataset.py:
```diff
@@ -138,6 +158,125 @@ def to_file(self, filename: Path, file_format: str = "xyz"):
+            forces = None
+
+        try:
+            stress = atoms.get_stress(voigt=True)
```
Added changes:
`utils` now has a `stress_to_voigt` function, to convert 3x3 numpy array from cauchy-stress colabfit to 1x6 list. So it is consist as of now. Can convert to 3x3 or add similar conversion function in future.
